### PR TITLE
Articulationpoints

### PR DESCRIPTION
**Pull request by jpfairbanks**
Improvements to #501 .

@sbromberger what do you think about making a type for holding the iteration state? I think that makes it easier to remember what the various components are. There is nothing wrong with the way it was written though (once you remove the global 😉).

### CODE REVIEW

File: src/biconnectivity/articulation.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            if state.low[w] >= state.depth[v] && u != v
+                state.articulation_points[v] = true
+            end
+
+        elseif w != u
+            state.low[v] = min(state.low[v], state.depth[w])
+        end
+	  end
+    if u == v && children > 1
+        state.articulation_points[v] = true
```
Comment (sbromberger) [inline src/biconnectivity/articulation.jl]: `state.articulation_points[v] = (u == v && children > 1)`, perhaps?

Comment (jpfairbanks) [inline src/biconnectivity/articulation.jl]: Oh yeah I like that.

File: src/biconnectivity/articulation.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
+"""
+Computes the articulation points(https://en.wikipedia.org/wiki/Biconnected_component)
+of a connected graph `g` and returns an array containing all cut vertices.
+"""
+function articulation(g::SimpleGraph) :: AbstractArray{Int}
+    state = Articulations(g)
+    for u in 1:nv(g)
```
Comment (sbromberger) [inline src/biconnectivity/articulation.jl]: `for u in vertices(g)`?

File: src/biconnectivity/articulation.jl [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    state.low[v] = state.depth[v]
+
+    for w in out_neighbors(g, v)
+        if state.depth[w] == 0
+            children += 1
+            visit!(state, g, v, w)
+
+            state.low[v] = min(state.low[v], state.low[w])
+            if state.low[w] >= state.depth[v] && u != v
+                state.articulation_points[v] = true
```
Comment (sbromberger) [inline src/biconnectivity/articulation.jl]: see below for the one-liner

**Comment by sbromberger** [comment]
This is pretty cool. I'm ok with the type for holding state.

**Comment by sbromberger** [comment]
@jpfairbanks let me know when you're ready to merge this....

**Comment by jpfairbanks** [comment]
I would like to get @athulappadan 's call on the `elif` without an `else` but after that I think it is good to go.

**Comment by jpfairbanks** [comment]
@sbromberger Your one liners don't work because you might visit the same index in the array more than once. Although they could be `array[v] |= (boolean formula)`

**Comment by jpfairbanks** [comment]
Tests pass && 🚢  it.

**Comment by sbromberger** [comment]
@jpfairbanks - not to hold this up, but I'm confused by this:
> Your one liners don't work because you might visit the same index in the array more than once

How does that affect this code? `state.articulation_points[v] = (u == v && children > 1)`

File: src/biconnectivity/articulation.jl [hunk 4]
```diff
@@ -0,0 +1,55 @@
+"""
+Computes the articulation points(https://en.wikipedia.org/wiki/Biconnected_component)
+of a connected graph `g` and returns an array containing all cut vertices.
+"""
+function articulation(g::SimpleGraph) :: AbstractArray{Int}
+    state = Articulations(g)
+    for u in vertices(g)
+        state.depth[u] == 0 && visit!(state, g, u, u)
```
Comment (sbromberger) [inline src/biconnectivity/articulation.jl]: I'm ok with this but we seem to be getting close to the line we drew a while back:

> Style point: prefer the short circuiting conditional over if/else when convenient, and where state is not explicitly being mutated (e.g., condition && error("message") is good; condition && i += 1 is not).

Comment (jpfairbanks) [inline src/biconnectivity/articulation.jl]: yeah I agree.

**Comment by jpfairbanks** [comment]
>> Your one liners don't work because you might visit the same index in the array more than once

> How does that affect this code? state.articulation_points[v] = (u == v && children > 1)

The first time we visit `v`, the one liner version is right, but the next time, the value in `state.articulation_points[v]` could be true but   `(u == v && children > 1)` evaluates to false. In which case, the if/then form would leave `state.articulation_points[v]` as true and the boolean assignment version would make it false.

**Comment by sbromberger** [comment]
> The first time we visit v, the one liner version is right, but the next time, the value in state.articulation_points[v] could be true but (u == v && children > 1) evaluates to false. 

Ah, of course. We don't set to `false` explicitly inside that block. OK. I got it.

Ready to merge?

**Comment by sbromberger** [comment]
@jpfairbanks if you're satisfied with #505, please feel free to close this one out and merge at your convenience.

**Comment by athulappadan** [comment]
@jpfairbanks The case Edge(1, 2) is visited then the elseif parts avoid visiting the Edge(2, 1).

**Pull request closed without merging**